pkg/trigger/trigger.go

```diff
@@ -53,23 +88,88 @@ type Trigger struct {
 
 	// closeChan is used to stop the background trigger routine
 	closeChan chan struct{}
+
+	triggerReasons *prometheus.CounterVec
+	triggerFolds   prometheus.Gauge
+	callDurations  *prometheus.HistogramVec
+
+	// numFolds is the current count of folds that happened into the
+	// currently scheduled trigger
+	numFolds int
+
+	// foldedReasons is the sum of all unique reasons folded together.
+	foldedReasons reasonStack
+
+	waitStart time.Time
 }
 
 // NewTrigger returns a new trigger based on the provided parameters
-func NewTrigger(p Parameters) *Trigger {
+func NewTrigger(p Parameters) (*Trigger, error) {
 	if p.sleepInterval == 0 {
 		p.sleepInterval = time.Second
 	}
 
+	if p.TriggerFunc == nil {
+		return nil, fmt.Errorf("trigger function is nil")
+	}
+
+	if p.PrometheusMetrics && p.Name == "" {
```

Comment:
It'd be nice to unit test these constraints, but doing that would require setting up Prometheus since it's not easily able to be mocked :(

Reply:
It's actually covered minimally in the DaemonTestSuite by setting up the metrics and incrementing them hence the change required to unregister the prometheus metric when a Daemon is no longer used. We are not testing whether the metrics are being incremented correctly though. I did that testing manually for now.